### PR TITLE
feat/11953: Support StringView for TRANSLATE() fn

### LINKED ISSUE #11953: Update `TRANSLATE` scalar function to support Utf8View
Part of https://github.com/apache/datafusion/issues/11752 and https://github.com/apache/datafusion/issues/11790

Currently, a call to `TRANSLATE` with a Utf8View datatypes induces a cast. After the change that fixes this issue, it should not.

```sql
query TT
EXPLAIN SELECT
  TRANSLATE(column1_utf8view, 'foo', 'bar') as c
FROM test;
----
logical_plan
01)Projection: translate(CAST(test.column1_utf8view AS Utf8), Utf8("foo"), Utf8("bar")) AS c
02)--TableScan: test projection=[column1_utf8view]
```

`translate` is defined here: https://github.com/apache/datafusion/blob/main/datafusion/functions/src/unicode/translate.rs

casting tests are in: https://github.com/apache/datafusion/blob/main/datafusion/sqllogictest/test_files/string_view.slt

### Is your feature request related to a problem or challenge?

We are working to add complete StringView support in DataFusion, which permits potentially much faster processing of string data. See https://github.com/apache/datafusion/issues/10918 for more background.

Today, most DataFusion string functions support DataType::Utf8 and DataType::LargeUtf8 and when called with a StringView argument DataFusion will cast the argument back to DataType::Utf8 which is expensive.

To realize the full speed of StringView, we need to ensure that all string functions support the DataType::Utf8View directly.

### Describe the solution you'd like

Update the function to support DataType::Utf8View directly

### Describe alternatives you've considered

The typical steps are:
1. Write some tests showing the function doesn't support Utf8View (see the tests in [`string_view.slt`](https://github.com/apache/datafusion/blob/main/datafusion/sqllogictest/test_files/string_view.slt) to ensure the arguments are not being cast
2. Change the `Signature` of the function to accept `Utf8View` in addition to `Utf8`/`LargeUtf8`
3. Update the implementation of the function to operate on `Utf8View`

Example PRs
* Update to use an arrow kernel that already supports StringView:  https://github.com/apache/datafusion/pull/11787
* Change the implementation to support StringView directly: https://github.com/apache/datafusion/pull/11676
* Change implementation (option 2): https://github.com/apache/datafusion/pull/11556

### Additional context

The documentation of string functions can be found here: https://datafusion.apache.org/user-guide/sql/scalar_functions.html#string-functions

To test a function with StringView with `datafusion-cli` you can use an example like this (replacing `starts_with` with the relevant function)
```sql
> create table foo as values (arrow_cast('foo', 'Utf8View'), arrow_cast('bar', 'Utf8View'));
0 row(s) fetched.
Elapsed 0.043 seconds.

> select starts_with(column1, column2) from foo;
+--------------------------------------+
| starts_with(foo.column1,foo.column2) |
+--------------------------------------+
| false                                |
+--------------------------------------+
1 row(s) fetched.
Elapsed 0.015 seconds.
```

To see if it is using utf8 view, use `EXPLAIN` to see the plan and verify there is no `CAST`. In this example the `CAST(column1@0 AS Utf8)` indicates that the function is not using `Utf8View` natively

```sql
> explain select starts_with(column1, column2) from foo;
+---------------+------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                         |
+---------------+------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | Projection: starts_with(CAST(foo.column1 AS Utf8), CAST(foo.column2 AS Utf8))                                                |
|               |   TableScan: foo projection=[column1, column2]                                                                               |
| physical_plan | ProjectionExec: expr=[starts_with(CAST(column1@0 AS Utf8), CAST(column2@1 AS Utf8)) as starts_with(foo.column1,foo.column2)] |
|               |   MemoryExec: partitions=1, partition_sizes=[1]                                                                              |
|               |                                                                                                                              |
+---------------+------------------------------------------------------------------------------------------------------------------------------+
2 row(s) fetched.
Elapsed 0.006 seconds.
```

It is also often good to test with a constant as well (likewise there should be no cast):

```sql
> explain select starts_with(column1, 'foo') from foo;
+---------------+----------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                     |
+---------------+----------------------------------------------------------------------------------------------------------+
| logical_plan  | Projection: starts_with(CAST(foo.column1 AS Utf8), Utf8("foo"))                                          |
|               |   TableScan: foo projection=[column1]                                                                    |
| physical_plan | ProjectionExec: expr=[starts_with(CAST(column1@0 AS Utf8), foo) as starts_with(foo.column1,Utf8("foo"))] |
|               |   MemoryExec: partitions=1, partition_sizes=[1]                                                          |
|               |                                                                                                          |
+---------------+----------------------------------------------------------------------------------------------------------+
2 row(s) fetched.
Elapsed 0.002 seconds.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by devanbenz**
## Which issue does this PR close?

Closes #11953 

## Rationale for this change

This will add StringView support for the TRANSLATE() function as part of the work to add complete StringView support in DataFusion, which permits potentially much faster processing of string data. See: https://github.com/apache/datafusion/issues/10918 for more background.

### CODE REVIEW

File: datafusion/sqllogictest/test_files/string_view.slt [hunk 1]
```diff
@@ -895,14 +895,13 @@ logical_plan
...
 ## Ensure no casts for TRANSLATE
-## TODO file ticket
 query TT
 EXPLAIN SELECT
   TRANSLATE(column1_utf8view, 'foo', 'bar') as c
 FROM test;
 ----
 logical_plan
-01)Projection: translate(CAST(test.column1_utf8view AS Utf8), Utf8("foo"), Utf8("bar")) AS c
+01)Projection: translate(test.column1_utf8view, Utf8("foo"), Utf8("bar")) AS c
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/string_view.slt:904]: I think we should probably add a test that actually invokes the function with a `StringVieAwArray` argument.

For example:

```
SELECT
  TRANSLATE(column1_utf8view, 'foo', 'bar') as c
```

File: datafusion/functions/src/unicode/translate.rs [hunk 2]
```diff
@@ -72,6 +75,7 @@ impl ScalarUDFImpl for TranslateFunc {
 
     fn invoke(&self, args: &[ColumnarValue]) -> Result<ColumnarValue> {
         match args[0].data_type() {
+            DataType::Utf8View => make_scalar_function(translate::<i32>, vec![])(args),
```
Comment (alamb) [inline datafusion/functions/src/unicode/translate.rs:78]: I suspect this I think you'll likely also need to change the implementation of `translate` as well (see for example what @PsiACE  has done in https://github.com/apache/datafusion/pull/11970)

**Review by alamb — COMMENTED** [review]
Thanks @devanbenz 


I tried this function out and it seems to get an internal error with now when trying to run with a StringArray column

```sql
DataFusion CLI v41.0.0
> create table foo as values (arrow_cast('foo', 'Utf8View'));
0 row(s) fetched.
Elapsed 0.027 seconds.

> SELECT  TRANSLATE(column1, 'foo', 'bar') from foo;
Internal error: could not cast value to arrow_array::array::byte_array::GenericByteArray<arrow_array::types::GenericStringType<i32>>.
This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker
```

File: datafusion/functions/src/unicode/translate.rs (hunk 2 above)
Comment (devanbenz) [reply datafusion/functions/src/unicode/translate.rs:78]: Sounds good I'll take a look 👍

**Comment by devanbenz** [comment]
@alamb I still need to implement the test case but the cast error should be gone now. I tried it out in the CLI (oops should have done that before! 😅) thanks for taking a look :) 

```
> create table foo as values (arrow_cast('foo', 'Utf8View'));
0 row(s) fetched.
Elapsed 0.026 seconds.

> SELECT  TRANSLATE(column1, 'foo', 'bar') from foo;
+------------------------------------------------+
| translate(foo.column1,Utf8("foo"),Utf8("bar")) |
+------------------------------------------------+
| foo                                            |
+------------------------------------------------+
1 row(s) fetched.
Elapsed 0.011 seconds.
```

File: datafusion/functions/src/unicode/translate.rs [hunk 3]
```diff
@@ -120,7 +144,7 @@ fn translate<T: OffsetSizeTrait>(args: &[ArrayRef]) -> Result<ArrayRef> {
             }
             _ => None,
         })
-        .collect::<GenericStringArray<T>>();
+        .collect::<StringArray>();
```
Comment (devanbenz) [inline datafusion/functions/src/unicode/translate.rs:147]: Will this cause an issue at all? To be honest I'm not entirely sure if there will be a perf issue by swapping out GenericStringArray and StringArray.

File: datafusion/sqllogictest/test_files/string_view.slt [hunk 4]
```diff
@@ -425,6 +425,19 @@ logical_plan
 01)Projection: starts_with(test.column1_utf8view, Utf8View("äöüß")) AS c1, starts_with(test.column1_utf8view, Utf8View("")) AS c2, starts_with(test.column1_utf8view, Utf8View(NULL)) AS c3, starts_with(Utf8View(NULL), test.column1_utf8view) AS c4
 02)--TableScan: test projection=[column1_utf8view]
 
+### Test TRANSLATE
+
+# Should run TRANSLATE using utf8view column successfully
```
Comment (devanbenz) [inline datafusion/sqllogictest/test_files/string_view.slt:430]: Hopefully this is enough, always happy to introduce additional test cases if necessary.

File: datafusion/functions/src/unicode/translate.rs (hunk 3 above)
Comment (devanbenz) [reply datafusion/functions/src/unicode/translate.rs]: Actually I will likely need to revert to GenericStringArray<T> so I can use i64 as a type on it by the looks StringArray is i32.

**Review by alamb — APPROVED** [review]
This is beautiful @devanbenz  -- thank you 🙏

**Pull request merged**